### PR TITLE
Include tooltip to Razor provisional completion

### LINKED ISSUE #7250: Provisional Completion in Razor should show full tooltip info
1. Open Visual Studio Code
2. Open a Razor file
3. Start completing `@DateTime.$$`

RESULT:
You can arrow up/down through the list, but the items don't have any tooltip to show their meaning:
image

EXPECT:
Tooltips with more data:
image
.... like in C# files.

### PR DESCRIPTION
**Pull request by jordi1215**
This resolves #7250. The cause was that when we added a provisional dot to the generated virtual csharp document, the change didn't get reflected on disk before the Roslyn server took the completion request.

Changes made in completionHandler.ts:
1. Made changes so that both provisional and non-provisional code completion uses the same `provideCompletionsCommand` (roslyn.provideCompletions).
2. On top of adding a temporary provisional dot "." after each provisional completion request, we are adding a "resolve" provisional dot for each resolve completion request. The dot is essential for the resolve request to be correct and therefor for the tooltip to show up.
3. Use vscode.workspace.onDidChangeTextDocument to ensure that the virtual documents get updated before sending the request to the roslyn server for both completion and resolve completion requests.

#### Before
image

#### What could have gone wrong if it wasn't for David's comment

https://github.com/user-attachments/assets/8147402c-a93e-4b59-aaff-fc682073bb8e

#### After
image

https://github.com/user-attachments/assets/534ca754-9cbe-4a03-9ba6-3b08c39416b2

### CODE REVIEW

**Comment by davidwengier** [comment]
> The documentation gets synced afterwards 

This is only if the completion is actually taken, right? Do things still work correctly otherwise? eg, type `DateTime.`, press ESC to dismiss completion, and then scroll the text into view and out of view, or hover over something etc. I would expect not cleaning up the `.` would cause disco colors etc.

**Comment by jordi1215** [comment]
> > The documentation gets synced afterwards
> 
> This is only if the completion is actually taken, right? Do things still work correctly otherwise? eg, type `DateTime.`, press ESC to dismiss completion, and then scroll the text into view and out of view, or hover over something etc. I would expect not cleaning up the `.` would cause disco colors etc.
> 
> TL;DR, If the Razor document doesn't change after typing the dot, presumably the generated document never gets re-synced



> > The documentation gets synced afterwards
> 
> This is only if the completion is actually taken, right? Do things still work correctly otherwise? eg, type `DateTime.`, press ESC to dismiss completion, and then scroll the text into view and out of view, or hover over something etc. I would expect not cleaning up the `.` would cause disco colors etc.
> 
> TL;DR, If the Razor document doesn't change after typing the dot, presumably the generated document never gets re-synced

Thanks for the heads up! I did manage to get it into a "bad" state. Assuming `var name="Jordi"`, `@name.` will trigger completion. If nothing gets accepted then C# will think it is an incomplete statement and give the red squiggles whereas HTML will interpret it as `Jordi.`. 

I think I will try add and remove the provisional dot for both completion and resolveCompletion command. 

Thanks so much!

File: src/razor/src/completion/completionHandler.ts [hunk 1]
```diff
@@ -217,54 +259,39 @@ export class CompletionHandler {
...
-            if (virtualDocument.removeProvisionalDot()) {
-                await this.ensureProjectedCSharpDocumentUpdated(virtualDocument.uri);
-            }
-        }
+    private async delay(ms: number) {
+        return new Promise((resolve) => setTimeout(resolve, ms));
     }
 
-    private async ensureProjectedCSharpDocumentUpdated(virtualDocumentUri: vscode.Uri) {
+    private async ensureProvisionalDotUpdatedInCSharpDocument(
```
Comment (ryzngard) [inline src/razor/src/completion/completionHandler.ts:276]: I wonder if it would be better to hook to `vscode.workspace.onDidChangeTextDocument` to wait for the buffer to be updated. Assuming we sync all changes before artifically adding the `.`

File: src/razor/src/csharp/csharpProjectedDocument.ts [hunk 2]
```diff
@@ -80,6 +82,7 @@ export class CSharpProjectedDocument implements IProjectedDocument {
         if (this.provisionalEditAt && this.preProvisionalContent) {
             // Undo provisional edit if one was applied.
             this.setContent(this.preProvisionalContent);
+            this.resolveProvisionalEditAt = this.provisionalEditAt;
```
Comment (davidwengier) [inline src/razor/src/csharp/csharpProjectedDocument.ts:85]: Perhaps this field should be cleared in `addProvisionalDotAt`, just to ensure the previous resolve index doesn't stick around?

File: src/razor/src/csharp/csharpProjectedDocument.ts [hunk 3]
```diff
@@ -88,6 +91,49 @@ export class CSharpProjectedDocument implements IProjectedDocument {
         return false;
     }
 
+    // add resolve provisional dot if a provisional completion request was made
+    // A resolve provisional dot is the same as a provisional dot, but it remembers the
+    // last provisional dot inserted location and is used for the roslyn.resolveCompletion API
+    public addResolveProvisionalDotAt() {
```
Comment (davidwengier) [inline src/razor/src/csharp/csharpProjectedDocument.ts:97]: Nit: I think `ensureResolveProvisionalDot` would be a better name for this. It doesn't always add, and on first reading of the handler that seemed like a potential bug, but it was just the name confusing me. Also drop the `At` since it doesn't take the index, it already knows.

**Review by davidwengier — COMMENTED** [review]
I agree with Andrew, it would be much better if we can hook the didChange rather than use a delay. The rest makes sense, though I must admit I find this whole area confusing and hard to reason about. Could just be my TypeScript skills :)

File: src/razor/src/completion/completionHandler.ts (hunk 1 above)
Comment (jordi1215) [reply src/razor/src/completion/completionHandler.ts:277]: This makes so much sense! The reflected update is on the new commit. Thanks!

File: src/razor/src/csharp/csharpProjectedDocument.ts (hunk 2 above)
Comment (jordi1215) [reply src/razor/src/csharp/csharpProjectedDocument.ts:89]: The problem here is right now the resolve completion request doesn't know whether the completion request was provisional or not. So one bad state could be: 
-> (Provisional completion) user triggers provisional completion
-> (resolve completion) provisional dot are added and removed as the user toggles completion list 
-> (regular completion) as the user continues to type, completion list shows up 
-> (resolve completion) provisional dot are added and removed still since the field is not cleared

File: src/razor/src/csharp/csharpProjectedDocument.ts (hunk 3 above)
Comment (jordi1215) [reply src/razor/src/csharp/csharpProjectedDocument.ts]: That makes sense! Thank you!

**Review by davidwengier — APPROVED** [review]

**Review by ryzngard — APPROVED** [review]
